### PR TITLE
Adding "where_clause" to filters

### PR DESCRIPTION
**Pull request by kavyasrinet**
# Description

This PR adds a new key `"where_clause"` to `"filters"` in the logical forms.  `"where_clause"` has an output type `dict` and can have keys : `"AND"/ "OR "/ "NOT"`. This change helps us move closer to a sql like structure and cleans up the filters. 

This change will move : `"triples"` to be inside `"where_clause"` (no more "triples" key), `"author"` to be a comparator (no more "author" key) and all "comparator" to be inside `"where_clause"` (no more "comparator" key in filters).

```
"filters" : {
    .....
    "where_clause" : {
         "AND": [<comparator> / <triples> / a list],
         "OR": [<comparator> / <triples> / a list],
         "NOT": [<comparator> / <triples> / a list],
        }
}
```

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [x] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [ ] I want a thorough review of the implementation.
- [x] I want a high level review. 
- [ ] I want a deep design review.

## Before and After
Changeset here: https://www.internalfb.com/phabricator/paste/view/P432653172

`"how many beds do you see"`

BEFORE...

```
{'dialogue_type': 'GET_MEMORY',
 'filters': {'output': 'COUNT',
             'triples': [{'obj_text': [0, [2, 2]], 'pred_text': 'has_name'}]}}
```

AFTER...
```
{'dialogue_type': 'GET_MEMORY',
 'filters': {'output': 'COUNT',
             'where_clause': {'AND': [{'obj_text': [0, [2, 2]],
                                       'pred_text': 'has_name'}]}}}
````
# Testing

Tested on devfair, locally and CI

# Checklist:

- [ ] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.

### CODE REVIEW

**Comment by aszlam** [comment]
most things work now.  the remaining things broken are in put memory- can you check that the format is right? e.g. in all_test_commands the put_memory have upsert clauses with "AND".  They should just have a triple, and where_clause only goes in the filters sub_dict (the one used for searching which memory is going to be updated).  Can you take a look and check to see if codemod is doing this?  I can just change the one in all_test_commands by hand if so

**Review by kavyasrinet — COMMENTED** [review]
Overall the changes on interpreter LGTM. Can you add some tests for the new functions defined in the changes ?
I'll take a look at the comment about Put_memory in test and dataset

**Comment by kavyasrinet** [comment]
@aszlam confirming that here's what `PUT_MEMORY` looks like in the dataset:

some examples:
```
can you please name this as very little
{"dialogue_type": "PUT_MEMORY", 
"filters": {"contains_coreference": "yes"}, 
"upsert": {
       "memory_data": {"memory_type": "TRIPLE", 
       "triples": [{"pred_text": "has_size", "obj_text": [0, [6, 7]]}]}}}


that silverfish is tall
{"dialogue_type": "PUT_MEMORY", 
"filters": {"selector": 
             {"location": {"reference_object": {
                   "special_reference": {"fixed_value": "SPEAKER_LOOK"}}}},
                  "where_clause": {"AND": [{"pred_text": "has_name", "obj_text": [0, [1, 1]]}]}}, "upsert": {"memory_data": {"memory_type": "TRIPLE", "triples": [{"pred_text": "has_tag", "obj_text": [0, [3, 3]]}]}}}

tag that stronghold
{"dialogue_type": "PUT_MEMORY", 
"filters": {"contains_coreference": "yes", 
          "where_clause": {"AND": [{"pred_text": "has_name", "obj_text": [0, [2, 2]]}]}}}

tag all items for destruction
{"dialogue_type": "PUT_MEMORY", 
"filters": {
      "where_clause": {"AND": [{"pred_text": "has_name", "obj_text": [0, [1, 2]]}]}}, 
"upsert": {"memory_data": {"memory_type": "TRIPLE", "triples": [{"pred_text": "has_tag", "obj_text": [0, [4, 4]]}]}}}
```

File: droidlet/interpreter/attribute_helper.py [hunk 1]
```diff
@@ -213,6 +213,4 @@ def __call__(self, interpreter, speaker, d_attribute, get_all=False):
     "born_time": "create_time",
     "modify_time": "updated_time",
     "visit_time": "attended_time",  # FIXME!!
-    "run_count": "run_count"
```
[the remark this answers is not in the document]
Comment (aszlam) [reply droidlet/interpreter/attribute_helper.py:216]: anything that stays the same isn't being canonicalized, compare https://github.com/facebookresearch/droidlet/blob/filters_update/droidlet/interpreter/attribute_helper.py#L198 to previous

File: droidlet/interpreter/craftassist/schematic_helper.py [hunk 2]
```diff
@@ -172,24 +196,29 @@ def interpret_named_schematic(
 
     warning:  if multiple possibilities are given for the same tag, current
     heursitic just picks one.  e.g. if the lf is
-        "triples" : [{"pred_text": "has_colour", "obj_text": "red"},
-                     {"pred_text": "has_colour", "obj_text": "blue"}]
+        "where_clause" : [{"pred_text": "has_colour", "obj_text": "red"},
```
[the remark this answers is not in the document]
Comment (aszlam) [reply droidlet/interpreter/craftassist/schematic_helper.py:199]: yeah, you are right

**Comment by aszlam** [comment]
> @aszlam confirming that here's what `PUT_MEMORY` looks like in the dataset:
> 
> some examples:
> 
> ```
> can you please name this as very little
> {"dialogue_type": "PUT_MEMORY", 
> "filters": {"contains_coreference": "yes"}, 
> "upsert": {
>        "memory_data": {"memory_type": "TRIPLE", 
>        "triples": [{"pred_text": "has_size", "obj_text": [0, [6, 7]]}]}}}
> 
> 
> that silverfish is tall
> {"dialogue_type": "PUT_MEMORY", 
> "filters": {"selector": {"location": {"reference_object": {
>                    "special_reference": {"fixed_value": "SPEAKER_LOOK"}}}},
>           "where_clause": {"AND": [
>                        {"pred_text": "has_name", "obj_text": [0, [1, 1]]}]}}, 
> "upsert": {"memory_data": {"memory_type": "TRIPLE", "triples": [{"pred_text": "has_tag", "obj_text": [0, [3, 3]]}]}}}
> 
> tag that stronghold
> {"dialogue_type": "PUT_MEMORY", 
> "filters": {"contains_coreference": "yes", 
>           "where_clause": {"AND": [{"pred_text": "has_name", "obj_text": [0, [2, 2]]}]}}}
> 
> tag all items for destruction
> {"dialogue_type": "PUT_MEMORY", 
> "filters": {
>       "where_clause": {"AND": [{"pred_text": "has_name", "obj_text": [0, [1, 2]]}]}}, 
> "upsert": {"memory_data": {"memory_type": "TRIPLE", "triples": [{"pred_text": "has_tag", "obj_text": [0, [4, 4]]}]}}}
> ```
> 
> So it looks like that was a manual error in tests let me fix it and re-run CI

lol that last one is wrong, probably from original annotation ("tag all items for destruction").... its not tag the thing that currently has the tag "all items" as "destruction"  :)

but the codemod/format does look correct to me

**Review by aszlam — APPROVED** [review]

File: droidlet/interpreter/craftassist/schematic_helper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aszlam) [reply droidlet/interpreter/craftassist/schematic_helper.py]: I just got rid of the comment altogether

**Pull request merged**